modules/userdirectory-canvas/src/main/java/org/opencastproject/userdirectory/canvas/CanvasUserProviderInstance.java

```diff
@@ -0,0 +1,708 @@
+
+    // Update cache statistics
+    canvasLoads.incrementAndGet();
+
+    Thread currentThread = Thread.currentThread();
+    ClassLoader originalClassloader = currentThread.getContextClassLoader();
+    try {
+      // Canvas userId (internal id), email address and display name
+      String[] canvasUser = getCanvasUser(userId);
+
+      if (canvasUser == null) {
+        // User not known to this provider
+        logger.debug("User {} not found in Canvas system", userId);
+        cache.put(userId, nullToken);
+        return null;
+      }
+
+      String email = canvasUser[0];
+      String displayName = canvasUser[1];
+
+      // Get the set of Canvas roles for the user
+      List<String> canvasRoles = getRolesFromCanvas(userId);
+
+      // If Canvas doesn't know about this user we need to return
+      if (canvasRoles == null || canvasRoles.isEmpty()) {
+        cache.put(userId, nullToken);
+        return null;
+      }
+
+      logger.debug("Canvas roles for user {}: {}", userId, canvasRoles.toString());
+
+      Set<JaxbRole> roles = new HashSet<>();
+
+      boolean isInstructor = false;
+
+      for (String r : canvasRoles) {
+        roles.add(new JaxbRole(r, jaxbOrganization, "Canvas external role", Role.Type.EXTERNAL));
+
+        if (r.endsWith(LTI_INSTRUCTOR_ROLE))
+          isInstructor = true;
+      }
+
+      // Group role for all Canvas users
+      roles.add(new JaxbRole(Group.ROLE_PREFIX + "CANVAS", jaxbOrganization, "Canvas Users", Role.Type.EXTERNAL_GROUP));
+
+      // Group role for Canvas users who are an instructor in one more courses
+      if (isInstructor)
+        roles.add(new JaxbRole(Group.ROLE_PREFIX + "CANVAS_INSTRUCTOR", jaxbOrganization, "Canvas Instructors",
+            Role.Type.EXTERNAL_GROUP));
+
+      logger.debug("Returning JaxbRoles: " + roles);
+
+      // JaxbUser(String userId, String password, String name, String email, String provider, boolean canLogin,
+      // JaxbOrganization organization, Set<JaxbRole> roles)
+      User user = new JaxbUser(userId, null, displayName, email, PROVIDER_NAME, jaxbOrganization, roles);
+
+      cache.put(userId, user);
+      logger.debug("Returning user {}", userId);
+
+      return user;
+
+    } finally {
+      currentThread.setContextClassLoader(originalClassloader);
+    }
+
+  }
+
+  /**
+   * Verify that the user exists
+   * Query with /api/v1/users/sis_user_id:{userId}
+   *
+   * @param userId Canvas User ID
+   */
+  private boolean verifyCanvasUser(String userId) {
+    logger.debug("verifyCanvasUser({})", userId);
+    try {
+      if ((userPattern != null) && !userId.matches(userPattern)) {
+        logger.debug("Verify user {} failed regexp {}", userId, userPattern);
+        return false;
+      }
+    } catch (PatternSyntaxException e) {
+      logger.warn("Invalid regular expression for user pattern {} - disabling checks, exception: {}", userPattern, e);
```

Comment:
For the logging, I suggest
```suggestion
      logger.warn("Invalid regular expression for user pattern {} - disabling checks", userPattern, e);
```

…but really, I'm wondering if we shouldn't completely fail here. Wat's the use-case for this fall-back? If it's a misconfiguration, it should get fixed, shouldn't it? And accidentally ignoring this since people don't notice that they have a problem because of this fallback can lead to unexpected results and may even be harmful, can't it?

Reply:
Agreed, I'll throw an exception here

Reply:
Hmm, looking into this more, essentially this would be a duplicate of the ConfigurationException written [here](https://github.com/SussexLearningSystems/opencast/blob/dcadeb6aa906c68721760b111e459ade8a1f7f74/modules/userdirectory-canvas/src/main/java/org/opencastproject/userdirectory/canvas/CanvasUserProviderFactory.java#L161-L168)?

I could go ahead and make this exception in that style so that it's clear what the exception is about, but it doesn't seem DRY to me.